hibernate-core/src/main/java/org/hibernate/bytecode/enhance/internal/bytebuddy/EnhancerImpl.java

```diff
@@ -165,6 +165,15 @@ public void discoverTypes(String className, byte[] originalBytes) {
 	}
 
 	private DynamicType.Builder<?> doEnhance(Supplier<DynamicType.Builder<?>> builderSupplier, TypeDescription managedCtClass) {
+		// skip if the class was already enhanced. This is very common in WildFly as classloading is highly concurrent.
+		// We need to ensure that no class is instrumented multiple times as that might result in incorrect bytecode.
+		// N.B. there is a second check below using a different approach: checking for the marker interfaces; I'm currently
+		// inclined to keep both checks for better backwards compatibility.
+		if ( managedCtClass.getDeclaredAnnotations().isAnnotationPresent( EnhancementInfo.class ) ) {
```

Comment:
Looks like a duplicate. A few lines below it already does that check i.e. see `if ( alreadyEnhanced( managedCtClass ) ) {`

Reply:
Sure but I did explain this in the comment :)

Reply:
My point is, the if-branch below will never be hit AFAIU because we add the `@EnhancementInfo` also to managed types, so the code below can probably just be removed.

Reply:
I think that the if-branch check below is a double-check that gives us a little more assurance that null is returned if a different thread already enhanced the class.  

Perhaps a third check could be done at the very tail end of this method which would give us more assurance that we will return null if another thread has already enhanced the class to ensure we don't add the EnhancementInfo more than once.  I suppose we could just move the second check to the tail-end as well as another option.  This is already marked resolved but thought I would add my 2 cents still... :-)

Reply:
the two checks are substantially different, one is based on the `Managed` interface being found in the inheritance tree - and does not require the `EnhancementInfo` at all, the other one skips based on the `EnhancementInfo`.
More importantly, the `Managed` interface could have been added manually or relate to legacy enhancement, hence my note about backwards compatibility.

I did wonder if I should have patched the `alreadyEnhanced` method rather than adding a separate block, but we're getting lost in the weeds :)